### PR TITLE
Upgrade libgit2 to 1.3.0

### PR DESCRIPTION
**Pull request by nalimilan**
Goes with https://github.com/JuliaLang/julia/pull/43250.

### CODE REVIEW

File: L/LibGit2/bundled/patches/libgit2-hostkey.patch
Comment (nalimilan) [inline L/LibGit2/bundled/patches/libgit2-hostkey.patch]: @StefanKarpinski This part of the patch seems to have been replaced by https://github.com/libgit2/libgit2/commit/29fe5f61ac2d5f7039ecf7dad834918fc683c98e. I'm not sure whether it's perfectly equivalent though. x-ref https://github.com/JuliaPackaging/Yggdrasil/pull/2419.

File: L/LibGit2/bundled/patches/libgit2-hostkey.patch
Comment (nalimilan) [inline L/LibGit2/bundled/patches/libgit2-hostkey.patch:44]: @StefanKarpinski This part of the patch seems to have been replaced by https://github.com/libgit2/libgit2/commit/29fe5f61ac2d5f7039ecf7dad834918fc683c98e. I'm not sure whether it's perfectly equivalent though. x-ref https://github.com/JuliaPackaging/Yggdrasil/pull/2419.

Comment (StefanKarpinski) [reply L/LibGit2/bundled/patches/libgit2-hostkey.patch:44]: Fun. I guess we need to make sure that the `certificate_callback` and `ssh_knownhost_check` stuff in `LibGit2` still works with the new libgit2 version. The patch was required, IIRC, because the old version of libgit2 was doing several broken things:

1. It was failing to pass the port number with the host, which is necessary when connecting over SSH using different ports and using the known hosts file.
2. It didn't pass the actual hostkey info from the connection to the callback, which made it impossible to do hostkey verification.

It's unclear how (or if) anyone was actually using SSH hostkey checking previously.

Comment (StefanKarpinski) [reply L/LibGit2/bundled/patches/libgit2-hostkey.patch:44]: It looks like the first thing is still broken but the second thing got fixed but in an incompatible way because they stuck that `raw_type` field in there instead of just appending the hostkey data.

Comment (nalimilan) [reply L/LibGit2/bundled/patches/libgit2-hostkey.patch:44]: Thanks. So AFAIK I kept the patch for the first thing so it should be OK. For the second one, tests added by https://github.com/JuliaLang/julia/pull/39324 cover it, right? At least they pass with libgit2 1.3 and this branch.

Comment (StefanKarpinski) [reply L/LibGit2/bundled/patches/libgit2-hostkey.patch:44]: I don't think that tests the port part but it looks like it tests the hostkey part. I was testing by actually connecting to a host that I was proxying through to github.

Comment (nalimilan) [reply L/LibGit2/bundled/patches/libgit2-hostkey.patch:44]: OK. I've just done this, and it seems to work!

For the record, here's what I did to forward port 22 on my machine to github.com:
```
sudo iptables -t nat -A PREROUTING -i wlo1 -p tcp --dport 32800  -j DNAT --to 38.140.156.5:22
sudo iptables -A FORWARD -i wlo1 -p tcp --dport 22 -d 38.140.156.5 -j ACCEPT
```
Then I had to go through my router as trying to connect from the same machine avoids the forwarding.

Then I installed a package via (after manually adding A.B.C.D to ~/.ssh/known_hosts with the signature for github.com):
```
]add ssh://git@A.B.C.D:32800/nalimilan/FreqTables.jl.git
```

One weird fact is that if I call `git clone ssh://git@A.B.C.D:32800/nalimilan/FreqTables.jl.git`, an entry for `[88.122.67.23]:32800` is added to ~/.ssh/known_hosts, but that entry isn't used by `]add ...` to work. I need to add the signature without the port. I guess that's expected since the same happens on Julia 1.6.

Anything I missed?

Comment (nalimilan) [reply L/LibGit2/bundled/patches/libgit2-hostkey.patch:44]: Bump.

**Comment by mkitti** [comment]
Would it make sense to pin this to MbedTLS 2.28.0?

**Comment by mkitti** [comment]
https://github.com/JuliaLang/julia/pull/42311 will not be able to succeed until we move the pin for LibGit2 to 2.28

```
LoadError("sysimg.jl", 19, LoadError("/buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.8/LibGit2/src/LibGit2.jl", 3, LoadError("/buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.8/LibGit2/src/utils.jl", 44, ErrorException("could not load library \"libgit2\"\nlibmbedtls.so.13: cannot open shared object file: No such file or directory"))))
```

**Comment by ViralBShah** [comment]
What do we need to get this merged?

**Comment by giordano** [comment]
Besides the fact it isn't clear to me whether the discussion above finished, we should decide whether to ship MbedTLS 2.28 in julia v1.8: if so, we must link libgit2 to that version (mbedtls breaks the ABI every time....)

**Comment by ViralBShah** [comment]
I do believe we want to ship MbedTLS 2.28 in julia v1.8, and because of the security vulnerabilities, we may even need to backport to 1.7 and 1.6

**Comment by giordano** [comment]
> we may even need to backport to 1.7 and 1.6

That's not going to work: since mbedtls breaks the ABI, it'll break any package depending on it, like libcurl, libgit2.

**Comment by nalimilan** [comment]
I think everything is ready, we just need a confirmation by @StefanKarpinski. https://github.com/JuliaLang/julia/pull/43250 bumps libgit2 in Julia and should be merged soon after this one.

**Comment by ViralBShah** [comment]
@nalimilan If things are good and passing tests, I suggest merging.

**Comment by ViralBShah** [comment]
I think we still need to pin to MbedTLS 2.28 before merging.

File: L/LibGit2/build_tarballs.jl
Comment (giordano) [inline L/LibGit2/build_tarballs.jl:61]:
```suggestion
    Dependency("MbedTLS_jll"; compat="~2.28.0"),
```

**Comment by giordano** [comment]
Note that you'll first need to update mbedtls 2.28 in julia, and perhaps update all dependent packages together at the same time, otherwise nothing will work at all.

**Comment by ViralBShah** [comment]
I assume that all Ygg PRs need to be merged first, and then Julia 1.8-dev does a single PR to bump them all. We still need an upstream release of libssh2.

**Comment by nalimilan** [comment]
@giordano Could you confirm what steps we should take?

**Comment by mkitti** [comment]
The issue to watch is https://github.com/JuliaPackaging/Yggdrasil/pull/4208 for libssh2. It would be preferable to have a version bump upstream, although we need to consider if we should just create a libssh2 `1.10.1` ourselves.

**Comment by giordano** [comment]
Yes, libgit2, libssh2 and the new mbedtls must all go together.

**Comment by nalimilan** [comment]
OK. TBH I don't really understand the situation with libssh2: is there anything preventing us from upgrading to 1.10.0 and handling the security fix after the feature freeze?

**Comment by ViralBShah** [comment]
We can certainly after the freeze. It is just a crucial CVE in mbedtls, so sooner the better.

**Comment by ViralBShah** [comment]
@giordano Are we good to merge this?

File: L/LibGit2/build_tarballs.jl
Comment (giordano) [inline L/LibGit2/build_tarballs.jl:62]: I think this is too restrictive: contrary to `MbedTLS`, I'm not aware of the fact libssh2 routinely breaks API/ABI in every minor version.  Good to have a compat, but I'd relax it to
```suggestion
    Dependency("LibSSH2_jll"; compat="1.10.1"),
```

**Review by giordano — APPROVED** [review]
This is ok with me.  Feel free to merge if there are no more concerns with compatibility on the julia side

**Comment by ViralBShah** [comment]
Only one way to find out.

**Pull request merged**
